controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -181,15 +183,21 @@ func (w *Workload) getCoreDNSInfo(ctx context.Context, clusterConfig *bootstrapv
 		return nil, err
 	}
 
+	// Handle the renaming of the upstream image from "k8s.gcr.io/coredns" to "k8s.gcr.io/coredns/coredns"
+	toImageName := parsedImage.Name
+	if toImageRepository == "k8s.gcr.io" && toImageName == "coredns" && targetMajorMinorPatch.GTE(semver.MustParse("1.8.0")) {
+		toImageName = "coredns/coredns"
+	}
```

Comment:
Should this be in a webhook rather than here? We also might want to add some validation

Reply:
This is not a property in any of our structs. We only have repository and tag, "coredns" or "coredns/coredns" is the part in between which is hard-coded in kubeadm right now

Reply:
What if a user specifies these values and then we apply something else without communicating it? That's why I was suggesting to have it in a validation webhook as well.

Reply:
The user can't specify the image name though right? only the repo and tag. The only case where the image name needs to be updated is if the user is _not_ specifying any custom values for coredns repo/tag and is using kubeadm defaults (k8s.gcr.io w/  v1.8.0 tag).

So if you're specifying something like:

```
ImageMeta: bootstrapv1.ImageMeta{
		// provide an newer image to update to
		ImageRepository: "myteam.azurecr.io",
		ImageTag:        "1.8.0",
},
```

which translates to `myteam.azurecr.io/coredns:1.8.0` there shouldn't be any updates.

Reply:
Exactly. The root cause which lead to this problem is that the properties in the struct do not fully specify the image name.
You get: `ImageRepository/CoreDNSImageName/ImageTag`. `CoreDNSImageName` is a hard-coded constant in kubeadm which changed from [1.20](https://github.com/kubernetes/kubernetes/blob/release-1.20/cmd/kubeadm/app/constants/constants.go#L332) to [1.21](https://github.com/kubernetes/kubernetes/blob/release-1.21/cmd/kubeadm/app/constants/constants.go#L331). I described it in full detail in https://github.com/kubernetes-sigs/cluster-api/issues/4463#issuecomment-819208692 (maybe a bit to verbose).

Reply:
What we're trying to achive in this PR is:
* if the user specified a non-default ImageRepostory => just change the tag during the deployment update
* if the user is using the default ImageRepository => handle the change from kubeadm 1.20 => 1.21 because otherwise we will just get an ImagePullBackoff (starting with ca. CoreDNS 1.8.0)

Reply:
The "full" solution to this problem would be to make the image name configurable in kubeadm and then use that in ClusterAPI. If we already had that, we could fix this via the webhook.

Reply:
Ah ok this makes sense. I misread while reviewing and thought we were modifying the image repository, not the name. This can be resolved.

Reply:
The image name is not visible in the API, so there is no way to mutate/validate inside web hooks